src/module/core-modules/napa/lock-wrap.cpp

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+#include "lock-wrap.h"
+
+#include <napa/module/binding/wraps.h>
+
+#include <memory>
+#include <mutex>
+#include <vector>
+
+using namespace napa::module;
+
+NAPA_DEFINE_PERSISTENT_CONSTRUCTOR(napa::module::LockWrap)
+
+void LockWrap::Init() {
+    auto isolate = v8::Isolate::GetCurrent();
+    auto constructorTemplate = v8::FunctionTemplate::New(isolate, DefaultConstructorCallback<LockWrap>);
+    constructorTemplate->SetClassName(v8_helpers::MakeV8String(isolate, exportName));
+    constructorTemplate->InstanceTemplate()->SetInternalFieldCount(1);
+
+    InitConstructorTemplate<LockWrap>(constructorTemplate);
+
+    NAPA_SET_PROTOTYPE_METHOD(constructorTemplate, "guardSync", GuardSyncCallback);
+
+    auto constructor = constructorTemplate->GetFunction();
+    InitConstructor("<LockWrap>", constructor);
+    NAPA_SET_PERSISTENT_CONSTRUCTOR(exportName, constructor);
+}
+
+v8::Local<v8::Object> LockWrap::NewInstance() {
+    return binding::CreateShareableWrap(std::make_shared<std::mutex>(), exportName);
+}
+
+void LockWrap::GuardSyncCallback(const v8::FunctionCallbackInfo<v8::Value>& args) {
+    auto isolate = v8::Isolate::GetCurrent();
+    v8::HandleScope scope(isolate);
+
+    int argc = args.Length();
+    CHECK_ARG(isolate, argc >= 1, "1 argument is required for calling 'guardSync'.");
+    CHECK_ARG(isolate, args[0]->IsFunction(), "Argument \"func\" shall be 'Function' type.");
+
+    auto context = isolate->GetCurrentContext();
+    auto holder = args.Holder();
+    auto thisObject = NAPA_OBJECTWRAP::Unwrap<LockWrap>(args.Holder());
+
+    v8::TryCatch tryCatch;
+    try {
+        auto mutex = thisObject->Get<std::mutex>();
+        std::lock_guard<std::mutex> guard(*mutex);
+
+        std::vector<v8::Local<v8::Value>> rest;
+        rest.reserve(argc - 1);
+
+        for (int i = 1; i < argc; i++) {
+            rest.emplace_back(args[i]);
+        }
```

Comment:
how about make parameters of func comes in an array. It can help you extend lock.guard with more parameters in future.